File: Homework 6/Sources/GTUMap.cpp
#include "../Headers/GTUMap.h"

namespace GTU
{
    template <class K, class V>
    GTUMap<K, V>::GTUMap(int width) : GTUSet<std::pair<K, V>>(width)
    {/*--.--*/}
    
    template <class K, class V>
    GTUMap<K, V>::GTUMap(const GTUSetBase<std::pair<K, V>> & other): GTUSet<std::pair<K, V>>(other)
    {/*--.--*/}
    
    template <class K, class V>
    GTUMap<K, V>::GTUMap(const typename GTUSetBase<std::pair<K, V>>::GTUIterator & first, const typename GTUSetBase<std::pair<K, V>>::GTUIterator & last) : GTUSet<std::pair<K, V>>(first, last)
    {/*--.--*/}

    template <class K, class V>
    void GTUMap<K, V>::insert(const std::pair<K, V> & data) throw(std::bad_pafram)
    {
        if(count(data) != 0)
            throw std::bad_pafram();
        GTUSet<std::pair<K, V>>::insert(data);
    }

    template <class K, class V>
    int GTUMap<K, V>::count(const std::pair<K, V> & data) throw(std::bad_pafram)
    {
        int i = 0;
        for(typename GTUSetBase<std::pair<K, V>>::GTUIterator k = this->begin(); k != this->end(); ++k)
            if(std::get<0>(*k) == std::get<0>(data) || std::get<1>(*k) == std::get<1>(data))
                ++i;
        return i;
    }

    template <class K, class V>
    const V & GTUMap<K, V>::operator [](const K & input) const noexcept
    {
        typename GTUSetBase<std::pair<K, V>>::GTUIterator x = this->begin();
        for(; x != this->end() && std::get<0>(*x) != input; ++x);
        return std::get<1>(*x);
    }

    template <class K, class V>
    V & GTUMap<K, V>::operator [] (const K & input) noexcept
    {
        typename GTUSetBase<std::pair<K, V>>::GTUIterator x = this->begin();
        for(; x != this->end() && std::get<0>(*x) != input; ++x);
        return std::get<1>(*x);
    }
}

File: Homework 7/151044062/151044062/151044062/src/GTUMap.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */

import java.security.InvalidParameterException;
import javafx.util.Pair;

/**
 *
 * @author Muaz Kurt, 151044062
 * @param <K> for the Key object type.
 * @param <V> for the Value object type.
 */
public class GTUMap<K, V> extends GTUSet<Pair<K, V>>
{
    public GTUMap()
    {
        super();
    }
    public GTUMap(int size)
    {
        super(size);
    }
    public GTUMap(GTUSetInt<Pair<K,V>> other)
    {
        super(other);
    }
    public GTUMap(GTUIterator<Pair<K, V>> start, GTUIterator<Pair<K, V>> end)
    {
        super(start, end);
    }
    
    /**
     * Takes the given key and searches it for this set. If there is no Key such
     * this input, then returns null.
     * @param k for the searching Key value.
     * @return V type object. The object is the object of given Key.
     * 
     */
    public V at(K k)
    {
        for(GTUIterator<Pair<K, V>> search = begin(); search.hasNext(); search.next())
            if(search.index().getKey() == k)
                return search.index().getValue();
        return null;
    }
    
    @Override
    public int count(Object input)
    {
        int count = 0;
        for(GTUIterator<Pair<K, V>> search = begin(); search.hasNext(); search.next())
            if(search.index().getKey().equals(((Pair<K, V>) input).getKey()))
                ++count;
        return count;
    }

    @Override
    @SuppressWarnings("empty-statement")
    public GTUIterator<Pair<K,V>> find(Object input)
    {
        GTUIterator<Pair<K, V>> search = begin();
        for(; !search.index().getKey().equals((K)input) && search.hasNext(); search.next());
        return search;
    }

    @Override
    @SuppressWarnings("empty-statement")
    public void erase(Object input)
    {
        GTUIterator<Pair<K, V>> search = begin();
        for(; !search.index().getKey().equals((K) input) && search.hasNext(); search.next());
        if(search.index().getKey().equals((K) input))
            super.erase(search.index());
    }

}